### PR TITLE
[RFC]Add support for graceful shutdown for twemcache/slimcache

### PR DESCRIPTION
**Pull request by michalbiesek**
- add support for  SIGTERM which should allow calling properly teardown
methods

---
This change is [Reviewable](https://reviewable.io/reviews/pmem/pelikan/32)

### CODE REVIEW

**Review by KFilipek — COMMENTED** [review]
> Reviewed 5 of 5 files at r1.
> *[Reviewable](https://reviewable.io/reviews/pmem/pelikan/32)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @jschmieg)

File: src/server/slimcache/main.c [hunk 1]
```diff
@@ -79,6 +88,18 @@ teardown(void)
     log_teardown();
 }
 
+static void
+_shutdown(int signo)
+{
+    log_stderr("_shutdown received signal %d", signo);
+    worker_processor.running = WORKER_STOP;
```
Comment (pbalcer) [inline src/server/slimcache/main.c]: read and write must be atomic.

File: src/server/ds/main.c [hunk 2]
```diff
@@ -17,6 +17,7 @@ struct data_processor worker_processor = {
     ds_process_read,
     ds_process_write,
     ds_process_error,
+    true
```
Comment (pbalcer) [inline src/server/ds/main.c]: .running = true

That was my original suggestion. Your choice though.

File: src/server/slimcache/main.c [hunk 3]
```diff
@@ -79,6 +88,18 @@ teardown(void)
     log_teardown();
 }
 
+static void
+_shutdown(int signo)
+{
+    log_stderr("_shutdown received signal %d", signo);
+    __atomic_store_n(&worker_processor.running, false, __ATOMIC_SEQ_CST);
```
Comment (pbalcer) [inline src/server/slimcache/main.c]: memory_order_release

Comment (michalbiesek) [reply src/server/slimcache/main.c]: Done via __ATOMIC_RELEASE

File: src/server/ds/main.c (hunk 2 above)
Comment (michalbiesek) [reply src/server/ds/main.c]: Done.

File: src/server/slimcache/main.c (hunk 1 above)
Comment (michalbiesek) [reply src/server/slimcache/main.c]: Done.

**Review by pbalcer — APPROVED** [review]

**Pull request merged**